src/stackmap_checker/call_stack_state.c

```diff
@@ -108,51 +108,64 @@ size_t get_locations(stack_map_t *sm, call_stack_state_t *state,
                      uint64_t **locs)
 {
     size_t num_locations = 0;
+    size_t loc_index = 0;
     size_t new_size = 0;
     // Each record corresponds to a stack frame.
     for (size_t i = 0; i < state->depth; ++i) {
         stack_map_record_t opt_rec = state->records[i];
         stack_map_record_t *unopt_rec =
             stmap_get_map_record(sm, ~opt_rec.patchpoint_id);
-        new_size += opt_rec.num_locations * sizeof(uint64_t);
+        num_locations += opt_rec.num_locations;
+        new_size = num_locations * sizeof(uint64_t);
         *locs = (uint64_t *)realloc(*locs, new_size);
 
         // Populate the stack of the optimized function with the values the
         // unoptimized function expects
-        for (int j = 0; j < unopt_rec->num_locations - 1; j += 2) {
+        for (size_t j = 0; j + 1 < unopt_rec->num_locations; j += 2) {
             location_type type = unopt_rec->locations[j].kind;
-            uint64_t opt_location_value =
-                stmap_get_location_value(sm, opt_rec.locations[j],
-                                         state->registers[i],
-                                         (void *)state->bps[i]);
+            void *opt_location_value = NULL;
+            uint64_t *loc_size = NULL;
 
-            uint64_t loc_size =
-                stmap_get_location_value(sm, opt_rec.locations[j + 1],
-                                         state->registers[i],
-                                         (void *)state->bps[i]);
-            if (type == DIRECT) {
-                memcpy(*locs + num_locations, &opt_location_value, loc_size);
-                ++num_locations;
-                memcpy(*locs + num_locations, &loc_size, sizeof(uint64_t));
-                ++num_locations;
-            } else if (type == REGISTER) {
-                memcpy(*locs + num_locations, &opt_location_value, loc_size);
-                ++num_locations;
-                memcpy(*locs + num_locations, &loc_size, sizeof(uint64_t));
-                ++num_locations;
+            stmap_get_location_value(sm, opt_rec.locations[j + 1],
```

Comment:
So why do we get an initial 64-bit value, and then another variable sized one?

Reply:
The 64-bit value represents the size of the location at index `j`. The pass which records the live locations also inserts their sizes into the stack map. The sizes are represented as `uint64_t`s, which is why the initial value has a fixed size. This was necessary because the stack map record only specifies the address of each location, and the size is unknown.

(Actually, the size of each location is automatically recorded in the stack map, but it is kind of useless, since it is always `8` [the size of a pointer]).